File: lecture05/04_sumitup/template.cpp
// sumitup.cpp

#include <algorithm>
#include <iostream>
#include <numeric>
#include <set>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

// Functions

int read_input(vector<int> &numbers) {
    int target, n;

    if (cin >> target >> n && !target) {
    	return target;
    }

    numbers.resize(n);
    for (int i = 0; i < n; i++) {
    	cin >> numbers[i];
    }

    sort(numbers.rbegin(), numbers.rend());
    return target;
}

void sumitup(vector<int> s, vector<int> &c, size_t k, int target, set<string> &r) {
    // TODO Base: Subset sums up to target, so display it

    // TODO Base: Prune search when index is size of candidates or when subset
    // sum is greater than target

    // TODO Recurse: Try without current number and then with it 
}

// Main execution

int main(int argc, char *argv[]) {
    int         target;
    vector<int> numbers, subset;

    while ((target = read_input(numbers))) {
        set<string> results;
    	sumitup(subset, numbers, 0, target, results);

    	cout << "Sums of " << target << ":" << endl;
    	if (results.empty()) {
    	    cout << "NONE" << endl;
	} else {
	    for (auto it = results.rbegin(); it != results.rend(); it++) {
	    	cout << *it << endl;
	    }
	}
    }
}

File: lecture05/04_sumitup/template.py
#!/usr/bin/env python3

import itertools
import sys

# Functions

def sumitup_r(subset, numbers, index, target):
    # TODO Base case: subset sums up to target, so yield it

    # TODO Base case: prune search when index is length of numbers, or sum of susbet
    # is greater than target

    # TODO Recursive step: try with current number and without

def sumitup(numbers, target):
    # TODO: Recursively generate sums, collect results into set, and return
    # sorted results

def main():
    for line in sys.stdin:
        data    = list(map(int, line.split()))
        target  = data[0]
        n       = data[1]
        numbers = data[2:]

        if target == 0:
            break

        print('Sums of {}:'.format(target))
        results = sumitup(numbers, target)
        if results:
            for result in results:
                print('+'.join(map(str, result)))
        else:
            print('NONE')

# Main execution

if __name__ == '__main__':
    main()
